lib/bap_future/bap_future.ml

```diff
@@ -0,0 +1,503 @@
+open Core_kernel.Std
+
+module Std = struct
+
+  module Future = struct
+    type 'a waiters = {
+      mutable waiters : ('a -> unit) list
+    }
+
+    type direct = [`full   | `wait ]
+    type linked = [ direct | `link ]
+
+    type 'a t = ('a,linked) future
+    and ('a,'s) future = { mutable cell : ('a,'s) cell }
+    and ('a,_) cell =
+      | Full : 'a         -> ('a,[> `full]) cell
+      | Wait : 'a waiters -> ('a,[> `wait]) cell
+      | Link : 'a t       -> ('a,[> `link]) cell
+
+    let is_decided = function
+      | {cell=Full _} -> true
+      | _ -> false
+
+    let rec unlink future : ('a,direct) cell = match future.cell with
+      | Full _ as c -> c
+      | Wait _ as c -> c
+      | Link f -> unlink f
+
+    let cell f = unlink f
+
+    let decide v x = match cell v with
+      | Full _ -> invalid_arg "future is already decided"
+      | Wait {waiters} ->
+        v.cell <- Full x;
+        List.iter waiters ~f:(fun f -> f x)
+
+    module P = struct
+      type nonrec 'a t = 'a t
+      let create p = p
+      let fulfill p x = decide p x     
+      let is_fulfilled p = is_decided p
+    end
+    
+    let full x = {cell = Full x}
+
+    let empty () = {cell = Wait {waiters=[]}}
+
+    let create () =
+      let future = empty () in
+      future, future
+
+    let upon v f = match cell v with
+      | Full x -> f x
+      | Wait w -> w.waiters <- f :: w.waiters
+
+    let rec link t t' =
+      match t.cell, t'.cell with
+      | Link t,_  -> link t t'
+      | _,Link t' -> link t t'
+      | _,Full s -> t.cell <- Full s
+      | Full x,_ -> decide t' x
+      | Wait wx, Wait wy ->
+        wx.waiters <- (wx.waiters @ wy.waiters);
+        t'.cell <- Link t
+
+    let bind v f : 'b t = match cell v with
+      | Full x -> f x
+      | Wait w ->
+        let u = empty () in
+        w.waiters <- (fun x -> link u (f x)) :: w.waiters;
+        u
+
+    let map v ~f = match cell v with
+      | Full x -> full (f x)
+      | Wait w ->
+        let u = empty () in
+        w.waiters <- (fun x -> link u (full (f x))) :: w.waiters;
+        u
+
+    let peek v = match cell v with
+      | Full x -> Some x
+      | Wait _ -> None
+
+    let peek_exn v = match cell v with
+      | Full x -> x
+      | Wait _ -> invalid_arg "peek_exn: empty future"
+
+    module Monad = Monad.Make(struct
+        type nonrec 'a t = 'a t
+        let bind = bind
+        let map = `Custom map
+        let return = full
+      end)
+
+    module App = Applicative.Of_monad(struct
+        type nonrec 'a t = 'a t
+        include Monad
+      end)
+
+    module Args = Applicative.Make_args(struct
+        type nonrec 'a t = 'a t
+        include App
+      end)
+    include App
+    include Monad
+  end
+
+  type 'a future = 'a Future.t
+
+  type 'a signal = Signal of ('a -> unit)
+
+  module Promise = Future.P
```

Comment:
that's ugly, don't intermix modules like that. It would be much easier to work with the `Promise` module if it will contain implementation of its own functions, instead of mixing everything in future, and then masking. This is the main idea of modular programming - the ability to pack code into structural units.